File: 2015/23/javascript/main.mjs
import fs from "node:fs";

// --- Constants ---
const INPUT_FILE = "../input/input.txt";

// --- Functions ---
function CPU() {
	let cpu = { a: 0, b: 0 };
	cpu.set = function(reg, value) {
		switch (reg) {
			case "a": this.a = value; break;
			case "b": this.b = value; break;
			default: throw "Illegal registry: " + reg;
		}
	};
	cpu.get = function(reg) {
		switch (reg) {
			case "a": return this.a;
			case "b": return this.b;
			default: throw "Illegal registry: " + reg;
		}
	};
	cpu.exec = function(code) {
		let i = 0;
		while (i < code.length) {
			let s = code[i].split(" ");
			s[1] = s[1].replace(",", "");
			switch (s[0]) {
				case "hlf": this.set(s[1], this.get(s[1]) / 2); i++; break;
				case "tpl": this.set(s[1], this.get(s[1]) * 3); i++; break;
				case "inc": this.set(s[1], this.get(s[1]) + 1); i++; break;
				case "jmp": i += Number(s[1]); break;
				case "jie": (this.get(s[1]) % 2 === 0) ? i += Number(s[2]) : i++; break;
				case "jio": (this.get(s[1]) === 1) ? i += Number(s[2]) : i++; break;
				default: throw "Illegal operation: " + s[0];
			}
		}
	};
	return cpu;
}

// --- Parse the input file ---
const input = fs.readFileSync(INPUT_FILE, 'utf8').trim();
const code = input.split("\n");

// --- Puzzle 1 ---
let cpu = CPU();
cpu.exec(code);
console.log("1. Register b: ", cpu.b.toLocaleString());

// --- Puzzle 2 ---
cpu = CPU();
cpu.a = 1;
cpu.exec(code);
console.log("2. Register b: ", cpu.b.toLocaleString());

File: 2015/23/go/main.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

// --- Constants ---
const FILENAME = "../input/input.txt"

// --- Structs ---
type CPU struct {
	a int
	b int
}

func (cpu *CPU) set(reg rune, i int) {
	switch reg {
	case 'a':
		cpu.a = i
	case 'b':
		cpu.b = i
	}
}

func (cpu *CPU) get(reg rune) int {
	switch reg {
	case 'a':
		return cpu.a
	case 'b':
		return cpu.b
	}
	return 0
}

func (cpu *CPU) exec(code []string) {
	for i := 0; i < len(code); i++ {
		s := strings.Fields(code[i])
		s[1] = strings.ReplaceAll(s[1], ",", "")
		reg := rune(s[1][0])
		switch s[0] {
		case "hlf":
			cpu.set(reg, cpu.get(reg)/2)
		case "tpl":
			cpu.set(reg, cpu.get(reg)*3)
		case "inc":
			cpu.set(reg, cpu.get(reg)+1)
		case "jmp":
			i += parseInt(s[1]) - 1
		case "jie":
			if cpu.get(reg)%2 == 0 {
				i += parseInt(s[2]) - 1
			}
		case "jio":
			if cpu.get(reg) == 1 {
				i += parseInt(s[2]) - 1
			}
		default:
			log.Panic("Unsupported op:", s[0])
		}
	}
}

func parseInt(s string) int {
	n, err := strconv.Atoi(s)
	if err != nil {
		log.Panic(err)
	}
	return n
}

func main() {
	// --- Variables ---
	code := []string{}

	// --- Read input file ---
	file, err := os.Open(FILENAME)
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	// --- Parse each line ---
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		code = append(code, line)
	}

	// --- Puzzle 1 ---
	cpu := CPU{}
	cpu.exec(code)
	fmt.Println("1. Register b:", cpu.b)

	// --- Puzze 2 ---
	cpu = CPU{a: 1, b: 0}
	cpu.exec(code)
	fmt.Println("2. Register b:", cpu.b)
}
